File: code/sorting/src/bead_sort/bead_sort.c
#include <stdio.h>
#include <stdlib.h>
// Part of Cosmos by OpenGenus Foundation
void bead_sort(int *a, int len)
{
	int i, j, max, sum;
	unsigned char *beads;
#	define BEAD(i, j) beads[i * max + j]
 
	for (i = 1, max = a[0]; i < len; i++)
		if (a[i] > max) max = a[i];
 
	beads = calloc(1, max * len);
 
	/* mark the beads */
	for (i = 0; i < len; i++)
		for (j = 0; j < a[i]; j++)
			BEAD(i, j) = 1;
 
	for (j = 0; j < max; j++) {
		/* count how many beads are on each post */
		for (sum = i = 0; i < len; i++) {
			sum += BEAD(i, j);
			BEAD(i, j) = 0;
		}
		/* mark bottom sum beads */
		for (i = len - sum; i < len; i++) BEAD(i, j) = 1;
	}
 
	for (i = 0; i < len; i++) {
		for (j = 0; j < max && BEAD(i, j); j++);
		a[i] = j;
	}
	free(beads);
}
 
int main()
{
	int i, x[] = {5, 3, 1, 7, 4, 1, 1, 20};
	int len = sizeof(x)/sizeof(x[0]);
 
	bead_sort(x, len);
	for (i = 0; i < len; i++)
		printf("%d\n", x[i]);
 
	return 0;
}

File: code/sorting/src/bead_sort/bead_sort.go
package main

import (
	"fmt"
)

// Bead Sort in Golang
// Part of Cosmos by OpenGenus Foundation

// BeadSort sorts the given slice
func BeadSort(a []int) {
	n := len(a) // find the length of the array

	// this is used to handle for the edge case where there is nothing to sort
	if n == 0 {
		return
	}

	m := a[0]

	// get the max value of the array
	for _, value := range a {
		if value > m {
			m = value
		}
	}

	// create the grid of n rows and m columns and mark the beads
	grid := make([][]bool, n)
	for i, v := range a {
		grid[i] = make([]bool, m)
		for j := 0; j < v; j++ {
			grid[i][j] = true
		}
	}

	for j := 0; j < m; j++ {
		// count how many beads are on each post
		sum := 0
		for i := 0; i < n; i++ {
			if !grid[i][j] {
				continue
			}
			grid[i][j] = false
			sum++
		}

		// Move beads down
		for i := n - sum; i < n; i++ {
			grid[i][j] = true
		}
	}
	
	// Put sorted values in array using beads
	for i := 0; i < n; i++ {
		var j int
		for j = 0; j < m && grid[i][j]; j++ {

		}

		a[i] = j
	}
}

func main() {
	randomSlice := []int{5, 2, 8, 9, 10, 1, 3}
	fmt.Println("Unsorted Slice: ", randomSlice)
	BeadSort(randomSlice)
	// should receive [ 1, 2, 3, 5, 8, 9, 10 ]
	fmt.Println("Sorted Slice: ", randomSlice)
}
